1783 Implement Interface And Inherit Class

rmadsen-ks opened the pull request:
Added support for multiple inheritance when inheriting from one base class and/or multiple interfaces.

Added a unit test verifying that it works with a simple class and interface. This unit test would previously have failed since multiple types are in the class super class list.

Close #1783

### Checklist

Check all those that are applicable and complete.

-   [X] Make sure to include one or more tests for your change
-   [X] If an enhancement PR, please create docs and at best an example
-   [X] Ensure you have signed the [.NET Foundation CLA](https://cla.dotnetfoundation.org/pythonnet/pythonnet)
-   [X] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [X] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,56 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
             {
-                return Exceptions.RaiseTypeError("cannot use multiple inheritance with managed classes");
+                var base_type2 = Runtime.PyTuple_GetItem(bases, (int)i);
+                var cb2 = (ClassBase) GetManagedObject(base_type2);
+                if (cb2 != null)
```
You can use the same pattern to check for `null` as in the original code to clear compiler warning

e.g. `if (GetManagedObject(base_type2) is ClassBase cb2) ...`

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,56 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
+                {
+                    if (cb2.type.Valid && cb2.type.Value.IsInterface)
+                        interfaces.Add(cb2.type.Value);
+                    else baseType.Add(cb2);
```
IMHO, it is better to enforce the order to be `BaseClass, IInterface1, IInterface2`, etc.

I believe the current code has a bug down below because it uses `base_type` which is not the `BaseClass` if the order is different.

rmadsen-ks replied:
I have only added this functionality for types which goes through  TypeManager.CreateSubType, so it should only affect that branch.

for code flows not invoking CreateSubType, the behavior willl be exactly the same as before.

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,59 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
             }
 
-            // If __assembly__ or __namespace__ are in the class dictionary then create
+            // If the base class has a parameterless constructor, or
```
> If the base class has a parameterless constructor

??? What is this for? If that's right, it seems breaking to me.

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -130,10 +143,12 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
                 }
             }
 
+            var base_type = Runtime.PyTuple_GetItem(bases, 0);
```
`bases` here can be empty since you removed check above.

rmadsen-ks replied:
How can bases be length 0? This should only be called when a C# type is subclassed, right? In this case, it should at least be length 1.

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,59 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
-            // That type must itself have a managed implementation. We check
-            // that by making sure its metatype is the CLR metatype.
+            // Extract interface types and base class types.
+            var interfaces = new List<Type>();
```
IMHO, repeating the same base interface multiple times should be an error.

And it might be, add a test.

rmadsen-ks replied:
Test added.

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,59 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
+            // Multiple inheritance is not supported, unless the other types are interfaces
+            if (baseTypes.Count > 1)
+            {
+                return Exceptions.RaiseTypeError("cannot use multiple inheritance with managed classes");
```
Since you collected multiple base types, perhaps it would be useful to list them in the error message in case user did not notice which ones are classes.

rmadsen-ks replied:
Ok, I'll include this.

lostmsu commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,59 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
+
+                if (GetManagedObject(baseTypeIt) is ClassBase classBaseIt)
+                {
+                    if (classBaseIt.type.Valid && classBaseIt.type.Value.IsInterface)
```
Invalid `ClassBase` instances should also raise an error regardless of anything else.

rmadsen-ks replied:
got it, adding exceptions.

lostmsu reviewed (changes requested):
Please, rebase before posting a new iteration.

P.S. sorry for the late response

rmadsen-ks commented on src/runtime/Types/MetaType.cs:
```diff
@@ -79,49 +82,59 @@ public static NewReference tp_new(BorrowedReference tp, BorrowedReference args,
+                    if (classBaseIt.type.Valid && classBaseIt.type.Value.IsInterface)
+                        interfaces.Add(classBaseIt.type.Value);
+                    else baseTypes.Add(classBaseIt);
+                }
```
Added exceptions here in the latest version.

rmadsen-ks commented:
@lostmsu,

Thanks for doing the review.

I have rebased and fixed things you mentioned.

Will you mark the conversations resolved where appropriate?

lostmsu commented:
Thank you for the contribution!
